### PR TITLE
remote: fix usb power hub detection on some devices

### PR DESCRIPTION
**Pull request by duc996**
Not all usb hubs can be identified via DEVTYPE and DRIVER. For example
Raspberry 2 has DEVTYPE=usb_device and DRIVER=usb. Use the bDeviceClass
attribute to detect usb hubs for sure.

### CODE REVIEW

File: labgrid/resource/udev.py [hunk 1]
```diff
@@ -336,10 +336,16 @@ class USBPowerPort(USBResource):
...
-    def __attrs_post_init__(self):
-        self.match['DEVTYPE'] = 'usb_interface'
-        self.match['DRIVER'] = 'hub'
-        super().__attrs_post_init__()
+    def filter_match(self, device):
+        try:
+            devclass = device.attributes.asstring('bDeviceClass')
+            if devclass != '09':
+                return False
+        except UnicodeDecodeError:
```
Comment (jluebbe) [inline labgrid/resource/udev.py]: Do UnicodeDecodeErrors actually happen here? Could you use asint() instead?

File: labgrid/resource/udev.py [hunk 2]
```diff
@@ -336,10 +336,16 @@ class USBPowerPort(USBResource):
...
+            pass
+        except KeyError:
```
Comment (jluebbe) [inline labgrid/resource/udev.py]: If the attribute is missing, we need to return false, otherwise all devices would match.

**Comment by jluebbe** [comment]
Does the RPi's hub actually support power switching? Do you kown why it's using a different kernel driver?

**Comment by jluebbe** [comment]
@duc996 I'm not yet convinced that we actually need this change. Some questions from the previous review are still open:

> Does the RPi's hub actually support power switching? Do you kown why it's using a different kernel driver?

**Comment by jluebbe** [comment]
This is a power-switchable hub connected to the RPi and not the internal hub, correct? Could you show the output of `lsusb -t` and your exporter's yaml file?

**Comment by duc996** [comment]
> This is a power-switchable hub connected to the RPi and not the internal hub, correct? Could you show the output of `lsusb -t` and your exporter's yaml file?
No, it's on the card. 
lsusb -t
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=dwc_otg/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/5p, 480M
        |__ Port 1: Dev 3, If 0, Class=Vendor Specific Class, Driver=smsc95xx, 480M
        |__ Port 2: Dev 126, If 0, Class=Human Interface Device, Driver=usbhid, 480M
        |__ Port 4: Dev 124, If 0, Class=Vendor Specific Class, Driver=pl2303, 12M

uhubctl
Current status for hub 1-1 [0424:9514, USB 2.00, 5 ports]
  Port 1: 0503 power highspeed enable connect [0424:ec00]
  Port 2: 0503 power highspeed enable connect [15a2:0080 Freescale SemiConductor Inc SE Blank 6ULL]
  Port 3: 0100 power
  Port 4: 0103 power enable connect [067b:2303]
  Port 5: 0100 power

**Comment by duc996** [comment]
After searching a lot around, I finally think my path is ok. My configuration is correct, udevadm info gives the information: DEVTYPE=usb_device and DRIVER=usb, so self.match in USBPowerPort can't work. And if the device class indicates it's a hub (9) it must be a hub.

**Comment by egrigore** [comment]
Hi,

Was this solved in any way? I am facing the same problem with a D-link H7 hub, which works with uhubctl but can't be matched as an USBPowerPort because it has DRIVER='usb'.

**uhubctl output:**
```
$ uhubctl
Current status for hub 2-4 [05e3:0608 USB2.0 Hub, USB 2.10, 4 ports]
  Port 1: 0103 power enable connect [10c4:ea70]
  Port 2: 0100 power
  Port 3: 0503 power highspeed enable connect [1fc9:0134 NXP      SemiConductors Inc SE Blank M845S]
  Port 4: 0503 power highspeed enable connect [05e3:0608 USB2.0 Hub, USB 2.10, 4 ports]
$ uhubctl -l 2-4 -p 3 -a cycle
Current status for hub 2-4 [05e3:0608 USB2.0 Hub, USB 2.10, 4 ports]
  Port 3: 0503 power highspeed enable connect [0525:a4a5]
Sent power off request
New status for hub 2-4 [05e3:0608 USB2.0 Hub, USB 2.10, 4 ports]
  Port 3: 0000 off
Current status for hub 2-4 [05e3:0608 USB2.0 Hub, USB 2.10, 4 ports]
  Port 3: 0000 off
Sent power on request
New status for hub 2-4 [05e3:0608 USB2.0 Hub, USB 2.10, 4 ports]
  Port 3: 0100 power
```
**how hub looks - it has DRIVER='usb'**

```
$ udevadm info /sys/bus/usb/devices/2-4
P: /devices/pci0000:00/0000:00:1d.7/usb2/2-4
N: bus/usb/002/088
E: BUSNUM=002
E: DEVNAME=/dev/bus/usb/002/088
E: DEVNUM=088
E: DEVPATH=/devices/pci0000:00/0000:00:1d.7/usb2/2-4
E: DEVTYPE=usb_device
E: DRIVER=usb
E: ID_BUS=usb
E: ID_FOR_SEAT=usb-pci-0000_00_1d_7-usb-0_4
E: ID_MODEL=USB2.0_Hub
E: ID_MODEL_ENC=USB2.0\x20Hub
E: ID_MODEL_FROM_DATABASE=Hub
E: ID_MODEL_ID=0608
E: ID_PATH=pci-0000:00:1d.7-usb-0:4
E: ID_PATH_TAG=pci-0000_00_1d_7-usb-0_4
E: ID_REVISION=6052
E: ID_SERIAL=05e3_USB2.0_Hub
E: ID_USB_INTERFACES=:090000:
E: ID_VENDOR=05e3
E: ID_VENDOR_ENC=05e3
E: ID_VENDOR_FROM_DATABASE=Genesys Logic, Inc.
E: ID_VENDOR_ID=05e3
E: MAJOR=189
E: MINOR=215
E: PRODUCT=5e3/608/6052
E: SUBSYSTEM=usb
E: TAGS=:seat:
E: TYPE=9/0/1
E: USEC_INITIALIZED=1070499778211

```

**How I configured it in exporter and it fails to match it:**

```
variant 1
    USBPowerPort:
        match:
            'sys_name': '2-4'
        index: 3

variant 2
    USBPowerPort:
        match:
            'ID_PATH': 'pci-0000:00:1d.7-usb-0:4'
        index: 3

```

**Comment by duc996** [comment]
> Was this solved in any way?

Hi, I have to use the proposed patch to get this working on RPi. I don't know why it's not merged.
I have added comments to the review, should I do anything else?

**Comment by egrigore** [comment]
I used locally your patch , and it worked for my case. I don't know about the merge - this file was changed several times since the PR was made, someone from labgrid can help us with it - @jluebbe  @Emantor

**Comment by Emantor** [comment]
> Hi,
> 
> Was this solved in any way? I am facing the same problem with a D-link H7 hub, which works with uhubctl but can't be matched as an USBPowerPort because it has DRIVER='usb'.
> 
> **how hub looks - it has DRIVER='usb'**
> 
> ```
> $ udevadm info /sys/bus/usb/devices/2-4
>```
What does the interface look like, i.e. the output of `$ udevadm info /sys/bus/usb/devices/2-4:1.0`?
On my computers this has the `DRIVER=hub`, while the path without `:1.0` has `DRIVER=usb`.

**Comment by egrigore** [comment]
Yes, you are right the interface is mapped as an hub:
```
$ udevadm info /sys/bus/usb/devices/1-1.1.4:1.0
P: /devices/pci0000:00/0000:00:1a.0/usb1/1-1/1-1.1/1-1.1.4/1-1.1.4:1.0
E: DEVPATH=/devices/pci0000:00/0000:00:1a.0/usb1/1-1/1-1.1/1-1.1.4/1-1.1.4:1.0
E: DEVTYPE=usb_interface
E: DRIVER=hub
E: ID_MODEL_FROM_DATABASE=Hub
E: ID_USB_CLASS_FROM_DATABASE=Hub
E: ID_USB_PROTOCOL_FROM_DATABASE=Single TT
E: ID_VENDOR_FROM_DATABASE=Genesys Logic, Inc.
E: INTERFACE=9/0/0
E: MODALIAS=usb:v05E3p0608d6052dc09dsc00dp01ic09isc00ip00in00
E: PRODUCT=5e3/608/6052
E: SUBSYSTEM=usb
E: TYPE=9/0/1
E: USEC_INITIALIZED=20589165847762
E: net.ifnames=0
```

**Pull request closed without merging**